File: visualstudio/DiffStream.VisualStudio/Core/DiffService.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DiffStream.VisualStudio.Core
{
    public sealed class DiffService
    {
        private const int MaxVisibleChanges = 120;

        public DiffResult Diff(string oldContent, string newContent)
        {
            oldContent = Normalize(oldContent);
            newContent = Normalize(newContent);
            if (oldContent == newContent)
            {
                return SummaryOnly("No line changes detected.");
            }

            var oldLines = Split(oldContent);
            var newLines = Split(newContent);
            var operations = BuildOperations(oldLines, newLines);
            var result = new DiffResult();
            var hunk = new DiffHunk { OldStart = 1, NewStart = 1 };
            var oldLine = 1;
            var newLine = 1;
            var visibleChanges = 0;

            foreach (var operation in operations)
            {
                if (operation.Type != DiffLineType.Unchanged)
                {
                    visibleChanges++;
                    if (visibleChanges > MaxVisibleChanges)
                    {
                        result.Truncated = true;
                        if (operation.Type == DiffLineType.Added)
                        {
                            result.Summary.Added++;
                            newLine++;
                        }
                        else
                        {
                            result.Summary.Removed++;
                            oldLine++;
                        }

                        continue;
                    }
                }

                var line = new DiffLine { Type = operation.Type, Content = operation.Content };
                if (operation.Type == DiffLineType.Added)
                {
                    line.NewLineNumber = newLine++;
                    result.Summary.Added++;
                }
                else if (operation.Type == DiffLineType.Removed)
                {
                    line.OldLineNumber = oldLine++;
                    result.Summary.Removed++;
                }
                else
                {
                    line.OldLineNumber = oldLine++;
                    line.NewLineNumber = newLine++;
                }

                hunk.Lines.Add(line);
            }

            hunk.OldLines = hunk.Lines.Count(line => line.Type != DiffLineType.Added);
            hunk.NewLines = hunk.Lines.Count(line => line.Type != DiffLineType.Removed);
            result.Hunks.Add(hunk);
            result.Summary.Modified = CountModifiedPairs(hunk.Lines);
            result.Summary.Message = BuildMessage(result.Summary, result.Truncated);
            return result;
        }

        public DiffResult Created(string content, bool showContent)
        {
            return showContent ? Diff(string.Empty, content) : SummaryOnly("File created.");
        }

        public DiffResult Deleted(string content, bool showContent)
        {
            return showContent ? Diff(content, string.Empty) : SummaryOnly("File deleted.");
        }

        private static List<DiffLine> BuildOperations(string[] oldLines, string[] newLines)
        {
            var table = new int[oldLines.Length + 1, newLines.Length + 1];
            for (var i = oldLines.Length - 1; i >= 0; i--)
            {
                for (var j = newLines.Length - 1; j >= 0; j--)
                {
                    table[i, j] = oldLines[i] == newLines[j]
                        ? table[i + 1, j + 1] + 1
                        : Math.Max(table[i + 1, j], table[i, j + 1]);
                }
            }

            var operations = new List<DiffLine>();
            var oldIndex = 0;
            var newIndex = 0;
            while (oldIndex < oldLines.Length && newIndex < newLines.Length)
            {
                if (oldLines[oldIndex] == newLines[newIndex])
                {
                    operations.Add(new DiffLine { Type = DiffLineType.Unchanged, Content = oldLines[oldIndex] });
                    oldIndex++;
                    newIndex++;
                }
                else if (table[oldIndex + 1, newIndex] >= table[oldIndex, newIndex + 1])
                {
                    operations.Add(new DiffLine { Type = DiffLineType.Removed, Content = oldLines[oldIndex++] });
                }
                else
                {
                    operations.Add(new DiffLine { Type = DiffLineType.Added, Content = newLines[newIndex++] });
                }
            }

            while (oldIndex < oldLines.Length)
            {
                operations.Add(new DiffLine { Type = DiffLineType.Removed, Content = oldLines[oldIndex++] });
            }

            while (newIndex < newLines.Length)
            {
                operations.Add(new DiffLine { Type = DiffLineType.Added, Content = newLines[newIndex++] });
            }

            return operations;
        }

        private static DiffResult SummaryOnly(string message)
        {
            return new DiffResult { Summary = new DiffSummary { Message = message } };
        }

        private static string Normalize(string content) => content.Replace("\r\n", "\n").Replace("\r", "\n");

        private static string[] Split(string content)
        {
            if (string.IsNullOrEmpty(content))
            {
                return Array.Empty<string>();
            }

            return Normalize(content).TrimEnd('\n').Split('\n');
        }

        private static int CountModifiedPairs(IEnumerable<DiffLine> lines)
        {
            var list = lines.ToList();
            var count = 0;
            for (var i = 0; i < list.Count - 1; i++)
            {
                if (list[i].Type == DiffLineType.Removed && list[i + 1].Type == DiffLineType.Added)
                {
                    count++;
                }
            }

            return count;
        }

        private static string BuildMessage(DiffSummary summary, bool truncated)
        {
            var parts = new List<string>();
            if (summary.Added > 0) parts.Add($"{summary.Added} added");
            if (summary.Removed > 0) parts.Add($"{summary.Removed} removed");
            if (summary.Modified > 0) parts.Add($"{summary.Modified} modified");
            var message = parts.Count == 0 ? "No line changes detected." : string.Join(", ", parts);
            return truncated ? $"{message}. Large file change detected; showing first {MaxVisibleChanges} changes." : message;
        }
    }
}

File: src/core/DiffService.ts
import { diffLines } from "diff";
import type { DiffHunk, DiffLine, DiffResult, DiffSummary } from "../types/DiffTypes";
import { normalizeLineEndings, splitLines } from "../utils/fileUtils";

export interface DiffOptions {
  maxChanges?: number;
  contextLines?: number;
}

const DEFAULT_MAX_CHANGES = 120;
const DEFAULT_CONTEXT_LINES = 2;

export class DiffService {
  diff(oldContent: string, newContent: string, options: DiffOptions = {}): DiffResult {
    const normalizedOld = normalizeLineEndings(oldContent);
    const normalizedNew = normalizeLineEndings(newContent);

    if (normalizedOld === normalizedNew) {
      return this.emptyResult("No line changes detected.");
    }

    const maxChanges = options.maxChanges ?? DEFAULT_MAX_CHANGES;
    const contextLines = options.contextLines ?? DEFAULT_CONTEXT_LINES;
    const parts = diffLines(normalizedOld, normalizedNew, { newlineIsToken: false });

    let oldLine = 1;
    let newLine = 1;
    let current: DiffLine[] = [];
    let oldStart = 1;
    let newStart = 1;
    let leadingContext: DiffLine[] = [];
    const hunks: DiffHunk[] = [];
    let added = 0;
    let removed = 0;
    let unchanged = 0;
    let visibleChanges = 0;
    let truncated = false;

    const flush = () => {
      if (current.length === 0) {
        return;
      }
      const oldLines = current.filter((line) => line.type !== "added").length;
      const newLines = current.filter((line) => line.type !== "removed").length;
      hunks.push({ oldStart, oldLines, newStart, newLines, lines: current });
      current = [];
    };

    for (const part of parts) {
      const lines = splitLines(part.value);
      if (part.added) {
        if (current.length === 0) {
          oldStart = oldLine;
          newStart = newLine;
          current.push(...leadingContext);
        }
        for (const content of lines) {
          if (visibleChanges >= maxChanges) {
            truncated = true;
            newLine++;
            added++;
            continue;
          }
          current.push({ type: "added", newLineNumber: newLine, content });
          newLine++;
          added++;
          visibleChanges++;
        }
        leadingContext = [];
        continue;
      }

      if (part.removed) {
        if (current.length === 0) {
          oldStart = oldLine;
          newStart = newLine;
          current.push(...leadingContext);
        }
        for (const content of lines) {
          if (visibleChanges >= maxChanges) {
            truncated = true;
            oldLine++;
            removed++;
            continue;
          }
          current.push({ type: "removed", oldLineNumber: oldLine, content });
          oldLine++;
          removed++;
          visibleChanges++;
        }
        leadingContext = [];
        continue;
      }

      unchanged += lines.length;
      if (current.length > 0) {
        const context = lines.slice(0, contextLines);
        for (const content of context) {
          current.push({ type: "unchanged", oldLineNumber: oldLine, newLineNumber: newLine, content });
          oldLine++;
          newLine++;
        }
        flush();

        const remaining = lines.slice(context.length);
        if (remaining.length > contextLines) {
          oldLine += remaining.length - contextLines;
          newLine += remaining.length - contextLines;
          leadingContext = remaining.slice(-contextLines).map((content, index) => ({
            type: "unchanged",
            oldLineNumber: oldLine + index,
            newLineNumber: newLine + index,
            content
          }));
          oldLine += contextLines;
          newLine += contextLines;
        } else {
          leadingContext = remaining.map((content, index) => ({
            type: "unchanged",
            oldLineNumber: oldLine + index,
            newLineNumber: newLine + index,
            content
          }));
          oldLine += remaining.length;
          newLine += remaining.length;
        }
      } else {
        if (lines.length > contextLines) {
          oldLine += lines.length - contextLines;
          newLine += lines.length - contextLines;
          leadingContext = lines.slice(-contextLines).map((content, index) => ({
            type: "unchanged",
            oldLineNumber: oldLine + index,
            newLineNumber: newLine + index,
            content
          }));
          oldLine += contextLines;
          newLine += contextLines;
        } else {
          leadingContext = lines.map((content, index) => ({
            type: "unchanged",
            oldLineNumber: oldLine + index,
            newLineNumber: newLine + index,
            content
          }));
          oldLine += lines.length;
          newLine += lines.length;
        }
      }
    }

    flush();

    const modified = this.countModifiedPairs(hunks);
    const summary: DiffSummary = {
      added,
      removed,
      modified,
      unchanged,
      totalChanges: added + removed,
      message: this.buildSummaryMessage(added, removed, modified, truncated, maxChanges)
    };

    return { summary, hunks, truncated };
  }

  diffCreated(content: string, showContent: boolean, options?: DiffOptions): DiffResult {
    return showContent ? this.diff("", content, options) : this.summaryOnly("File created.");
  }

  diffDeleted(content: string, showContent: boolean, options?: DiffOptions): DiffResult {
    return showContent ? this.diff(content, "", options) : this.summaryOnly("File deleted.");
  }

  private emptyResult(message: string): DiffResult {
    return {
      summary: { added: 0, removed: 0, modified: 0, unchanged: 0, totalChanges: 0, message },
      hunks: [],
      truncated: false
    };
  }

  private summaryOnly(message: string): DiffResult {
    return {
      summary: { added: 0, removed: 0, modified: 0, unchanged: 0, totalChanges: 0, message },
      hunks: [],
      truncated: false
    };
  }

  private countModifiedPairs(hunks: DiffHunk[]): number {
    let modified = 0;
    for (const hunk of hunks) {
      for (let index = 0; index < hunk.lines.length - 1; index++) {
        if (hunk.lines[index].type === "removed" && hunk.lines[index + 1].type === "added") {
          modified++;
        }
      }
    }
    return modified;
  }

  private buildSummaryMessage(added: number, removed: number, modified: number, truncated: boolean, maxChanges: number): string {
    const parts = [
      added > 0 ? `${added} added` : "",
      removed > 0 ? `${removed} removed` : "",
      modified > 0 ? `${modified} modified` : ""
    ].filter(Boolean);
    const base = parts.length > 0 ? parts.join(", ") : "No line changes detected";
    return truncated ? `${base}. Large file change detected. Showing first ${maxChanges} changes.` : base;
  }
}
